### ISSUE TITLE
Roadmap planning - USE CASES WANTED

### ISSUE THREAD

**Issue by Jeffail** [Jeffail/benthos#503]
TLDR: Please share your use cases here or send them privately to hello@benthos.dev, we will use these to help steer the roadmap towards Benthos version 4.

Hey everyone,

Benthos has been at major version 3 for over a year now, and I consider it to be a pretty cool achievement that given all the dope features added we've managed to keep both the Benthos config spec and APIs fully backwards compatible.

However, eventually it would be nice to cut a new major release and prune all of the dead weight that has accumulated during this time. In order to do that I want to be sure that we've fully thought through any potential breaking changes that could be bundled along with it, and that requires a roadmap.

Therefore I'm hoping to hear from you on how you use Benthos, what works well for you and what doesn't, and any features you'd hope to see eventually that would make your lives easier.

Some features that might come into play are:

- Improved plugin APIs (need more use cases to help shape this)
- HTTP poll APIs for the `dynamic` components as well as streams mode, allowing Benthos to pull configs
- Configuration file reloading
- Expanded logging, metrics and tracing options

Please let us know how you use (or intend to use) Benthos, either by posting here or getting in touch privately at hello@benthos.dev. You can also find us on the discord channel at https://discord.gg/6VaWjzP, more chat options can be found on the [community page](https://www.benthos.dev/community).

Once planning starts I'm going to be looking into organising public roadmap planning sessions.

**Comment by jedjones-uk** [comment]
TL;DR: All of those listed features would be great, please don't remove any sources and sinks. JSON config is great, we almost always prefer that. Benthos is amazing and critical, we are interested in helping adoption in the wider community and cross over to other stacks.

**Our use case and rough plans.**

We are currently implementing a supporting PHP library that will make benthos integrate easily with a PHP system with the goal of allowing benthos to acts as the entry point to a PHP service. This is achieved currently through a plugin which adds an exec sink (named CLI) to benthos. This sink calls the PHP CLI and kicks everything off from there.

The libraries current set up work with the idea that the whole of the benthos config is defined via factories/classes within the PHP application itself, and then provides a simple bootstrap config YAML which benthos uses when starting in streams mode. This file looks like this:
```
input:
  bloblang:
    count: 1
    mapping: 'root = {"status":"ready"}'

buffer:
  none: {}

output:
  type: cli
  plugin:
    command: /usr/bin/php
    args: [ '/var/www/projects/php-benthos/dev/bootstrap.php' ] //The code in bootstrap really is just a single static call to a bootstrap function
```

To this end, I think having a way to set benthos to poll for a config, including with a first start pull, would simplify our whole bootstrap process. The polling would give a better update/deployment story as well, as changes will be picked up without needing a trigger.

The reason for having the config in PHP itself is that when streams are defined, they can also include abstract sink definitions which provide metadata, and type hinting so the CLI handler can call the class that was type hinted, rather than a switch case or some other methods of resolving the PHP side handler. The configs are then pushed over the benthos API as JSON data, so for us JSON is our primary configuration method.

We also intend to later support a custom fastcgi sink plugin as this would allow another method of integrating with PHP and make for a better microservice story when integrating with other frameworks and their routing.

**Why create a PHP benthos integration?** 

I work in the IoT industry, and we have an existing PHP application which processes all of the data as it is sent from various sources to us. This application is essentially a pipeline with stages, with a simple JSON end point that implements some very crude queueing on Redis. This application currently cannot be scaled horizontally as the semaphore/flag method of ensuring sequential processing of data per device is to unreliable and subject to race conditions.

The solution to that is a better queueing system such as kafka and it's consumer groups, as this will allow for multi stage processing and enforcement of sequential data. PHP and kafka have a poor integration in my opinion as PHP is not designed for long lived processes. By using benthos we dramatically extend the capabilities of our stream processor application to receive from many more sources, and then to republish to many more sinks.

This has evolved our existing application from a IoT gateway to a core microservice integration bus which allows us to collect and merge data from our b2b partners and their partners into our applications. We think this is a combination of functionality that will be appealing to many many legacy and new build PHP applications.

**Comment by jackxbritton** [comment]
v4! Exciting!

> HTTP poll APIs for the dynamic components as well as streams mode, allowing Benthos to pull configs

This is something we'd personally benefit from, great to see it's on the list!

Something else we'd like to see are optimizations for dealing with many (hundreds or thousands) of dynamic outputs in a single pipeline. In our use case, each dynamic output processes a subset of messages from the input. Something like this, after some dynamic outputs have already been POSTed to the Benthos instance:

```yaml
input: ...

output:
  dynamic:
    outputs:
    - http_client: ...
      processors:
      - bloblang: root = if type != "a" { deleted() }
    - kafka: ...
      processors:
      - bloblang: root = if type != "a" { deleted() }
    - sqs: ...
      processors:
      - bloblang: root = if type != "b" { deleted() }
    ...
```

The overhead of the filtering processors grows linearly with the number of dynamic outputs.

A solution could be to group messages upstream, so a new dynamic output could subscribe to just the messages it's interested in. Note that `type` could be anything, not just `"a"` or `"b"`. Maybe this grouping of messages could be defined as a resource, with a `key: ${! json("type") }` parameter?

**Comment by stergiotis** [comment]
**Meta-Programming:**
Perhaps this is just me but a *mild* form of meta-programming would be awesome. For examples in the form of templates  stored in the ressources section that can be instanciated with parameters which can be used in config-read-time evaluated bloblang expressions.

**Comment by jedjones-uk** [comment]
I would also like to add a request for dynamic resources in streams mode, as most of our config is set at run time. I'd guess that is being considered since the docs note to open a ticket, but thought I'd add it to the thread of completeness.

**Comment by dilipkk-foyernet** [comment]
We want to execute periodic(based on cron expression) queries against database and results of query needs to be processed using serveral rules so benthos workflow, bloblang and other components fits really well for processing logic but then we want to see if the cron expression based input generation using custom commands such as calling API or DB queries is something can be added to v4 road map. We totally understand if this something not benthos is designed for.

**Comment by Jeffail** [comment]
Hey @dilipkk-foyernet, is your requirement for the cron schedule itself to be determined by API/DB queries? Or just the payload itself?

**Comment by dilipkk-foyernet** [comment]
@Jeffail, ideally we want to use benthos configuration to specify cron expression or interval and when the schedule is matched the check/query against DB or API is executed then results should be passed to the set of benthos processors.

**Comment by jedjones-uk** [comment]
Along the same lines as @dilipkk-foyernet we'd really find use for this feature as well. I would say it would be best if it was implemented as a repeatable feature of various inputs so it can be used on others such as bloblang, file, http client, basically anywhere you'd use an interval, or the stream might have an "end".

I would also suggest it implements the [quartz style cron](https://flylib.com/books/en/2.65.1/the_cron_expression_format.html#:~:text=One%20of%20the%20differences%20is,schedules%20only%20to%20the%20minute.&text=Quartz%20uses%20the%20cron%20expression,the%20job%20at%20schedule%20time.) rather than unix because it is very similar but more flexible and fine grained.

**Comment by Jeffail** [comment]
Cool, I've opened https://github.com/Jeffail/benthos/issues/505 for the cron stuff. We currently support this stuff to an extent with the [bloblang input](https://www.benthos.dev/docs/components/inputs/bloblang), which provides you with a trigger, and then you can trigger an `http`, `sql` or `cache` processor to get the data you need to start the pipeline, and expanding it to support cron should be pretty straight forward.

I've added a secondary goal to that ticket in order to look at wrapping any input with a cron schedule, this is more difficult as some inputs aren't really designed to block indefinitely during a pause, but it also isn't blocked by a V4 so we could implement this at any time.

**Comment by Jeffail** [comment]
@jackxbritton, I've opened https://github.com/Jeffail/benthos/issues/506 for the dynamic output stuff. There's a few different ways of improving performance there so needs a bit more time to think it through, but this is definitely something I think is worth leaning into.

**Comment by dilipkk-foyernet** [comment]
@Jeffail Is possible to define resources using [resource type](https://www.benthos.dev/docs/components/processors/resource) and then select them dynamically based on message params?
```
pipeline:
  processors:
    - resource: foo_proc

resources:
  processors:
    foo_proc:
      jmespath:
        query: foo
```
instead of writing a hardcoded string for **foo_proc**, if we can select from this value from the message params/field then it can greatly enhance the pipeline definitions, it is almost like a wrapper processor which process the message and then invokes another process.  When there is no processor for matched with message param then we need to throw an error message.

**Comment by Jeffail** [comment]
Hey @dilipkk-foyernet, not currently, there's no major limitation stopping us from doing this but it'll need thinking through a fair amount in case there are unwanted side effects. However, you can use a [`switch` processor](https://www.benthos.dev/docs/components/processors/switch) to explicitly select resources.

**Comment by dilipkk-foyernet** [comment]
@Jeffail Is it possible to add Database Change data capture events as stream input? This is very useful to us, to deliver notifications, execute some logic based on those events.

**Comment by Jeffail** [comment]
Hey @dilipkk-foyernet, I've repurposed https://github.com/Jeffail/benthos/issues/82 as a general CDC issue. It's not particularly blocked or otherwise impacted by V4 so I likely wouldn't include it in the roadmap, but it's something we can get started on right now.

**Comment by dilipkk-foyernet** [comment]
Thanks :)

**Comment by vnnw** [comment]
How about adding scheduled job runner while running Benthos.
The use case is while doing stream processing, we might want to periodically do some tasks, such as house-cleaning, aggregating etc. For example, I read raw message from Kafka and enrich, transform them, then sink them into Clickhouse. At last, I periodically query CH and refresh the results for BI application such as dashboards.

**Comment by zaiboot** [comment]
Hello there. 
Is there any output support for RDBMS? Like PostgreSQL
Not sure if that is already possible on Benthos 3. I did some digging on the outputs and some scenarios can be cover by using REDIS cache but we need to output to PGSQL at some point. Is that something doable with benthos? I checked https://github.com/Jeffail/benthos/blob/feature/plugin-example/lib/input/plugins/mysql.go#L48 but I get a 404. 

Thanks

**Comment by Jeffail** [comment]
Hey @zaiboot, there's a general purpose [`sql` processor](https://www.benthos.dev/docs/components/processors/sql) that allows you to query or insert rows.

**Comment by zaiboot** [comment]
@Jeffail another idea might be elastic as a source / input. We usually have a vault of kafka messages and in some scenarios would like to read as a source.

**Comment by Jeffail** [comment]
@zaiboot I've opened https://github.com/Jeffail/benthos/issues/532, there are a few different ways of introducing a consumer of ES, I think a processor might be the best fit but maybe warrants a bit of looking into and some discussion.

**Comment by miko** [comment]
Potential use case: I am considering benthos for reading syslog-generated files (would need #271), using bloblang for serious data manipulation (as it is more powerful than e.g. telegraf or promtail), and then writing it to influxdb ( #297), sql databases, and/or prometheus, so it could be viewed with grafana.

**Comment by codingconcepts** [comment]
Hey @Jeffail,

Here are a couple of features that I'd love to see in benthos:

**File watchers for csv, s3 etc**:

Rather than close the input once a given file has been read, it would be great if benthos could continue to read files that are dumped into a given directory as they arrive.

``` yaml
input:
  csv_watcher:
    paths: [/some/directory]
    prefixes: ["*.1", "*.2"]
    parse_header_row: true
    delimiter: ','
```

Or perhaps more extensible, a generic "watcher" input could be the generic way to configure a watcher, and pipe the results to something that can then treat the file like csv file or anything else.

**(S)FTP intput**:

Similar to the above, it would be useful for interacting with systems that expose data through (S)FTP if there was support for these inputs in benthos.

**Comment by Jeffail** [comment]
Hey @codingconcepts, for S3 I'd recommend publishing upload notification events to SQS, which benthos can consume (https://docs.aws.amazon.com/AmazonS3/latest/dev/ways-to-add-notification-config-to-bucket.html). There's also an open issue for file watching: https://github.com/Jeffail/benthos/issues/271

I've created an issue for SFTP support: https://github.com/Jeffail/benthos/issues/539

**Comment by gedw99** [comment]
I intend to use it like in a CQRS style system where the devs using the system all design using grpc / protocol buffers but the methods in the protobufs are tagged to give meaning to the benthos pipeline.

Mutations coming in from the clients are sent to a badger based event queue that is setup with CRDT replication. This acts as a buffer to benthos and guarantees the at least once semantics that benthos needs. I could use NATS but prefer the ease of badgerdb customization.  Benthos picks up the mutations and transforms the data into the many read only badgerdb database systems. It uses the pmtsgs in the protocol buffers to help it be driven to run what are often Simple CRUD operations but more often are pumping data transformations into many read only badgerdb databases effectively treating them as material used views.

 Devs can query the read only tables using grpc with sql directly passed in as a argument.
The query returns the Data but is set in the system as a subscription also. Badgerdb has the ability to detect changes using the Subscribe function and these are turned into CUD events and checked against all active and stored query subscriptions by parsing the sql. Because there are no inner joins using genji to do this is relatively easy. The cud events that are determined to match the sql parser are then stored in a per query event store. Essentially acting as stored procedures in a traditional dB. These are then pushed to all clients. The offset is recorded by the client so that offline clients can catchup if they become network partitioned. 

For a user of the system they use grpc and tag which methods are commands to go into the event stream and tag which are queries that are subscriptions. The standard grpc code generator generates the code in golang and other languages.
Any other semantics like telemetry or Workflows can be described to give benthos enough meta data to do most of what it needs to do but also allow customized user scripted benthos scripts to also be included - it might be a matter of naming the benthos script in the tag. 

A secondary idea is to use the fact that benthos can fully run inside a WASM vm to put some of the logic at the edge inside a browser. It’s useful because it allows a low code environment to have a pipeline for customization. These benthos scripts would be saved in the database and loaded into the appropriate GUI modules at runtime. Why do this ? Because I am building gui frameworks off grpc where most of it is built at runtime via reflection but need a point in the unidirectional flow where they can add any business logic they need. The intended users are not devs but business users or GUI developers with no backend skills.

So the grpc IDL is used to drive the backend and the front end and benthos is also used at both tiers to allow non developers to augment the system.

I could use NATS , Liftbridge to handle the durable dB and message aspects but I want the system to be able to run as a single binary thrown onto any server. It’s much more secure I believe it’s much more closed and all data is encrypted at rest thanks to badgerdb rotating encryption keys. 
Also badgerdb can run anywhere. In the Browser , on a mobile. So can benthos ( although no tested using tinygo yet ) and so it seems a happy marriage. 

If anyone is interested in this concept your most welcome to get involved as a user or developer or help document the concept and use cases that a self reflective system like this can be used for.

The scaling aspects are  handled by hashing the key space and sharding the system with a global control plane updating all nodes. A bit like a kubernettes sidecar and the meta data system that all google GCE systems update and make available to all modes in a cluster.

**Comment by gedw99** [comment]
Migrations are slipstreaming into the event event stream. These are essentially updates to benthos scripts, protocol buffers and genji / badger dB read only stores .

Forgot that useful bit. It’s always too easy to forget how you will upgrade a system !!

**Comment by uknow2009** [comment]
Hey @Jeffail, I am new user to this awesome component. We have scenario to pull data from RDBMS (sqlserver/oracle/mysql) systems and put into destinations(which are already there in Benthos implementation). I was checking multiple docs/issues/pr and found below

> Hey @zaiboot, there's a general purpose [`sql` processor](https://www.benthos.dev/docs/components/processors/sql) that allows you to query or insert rows.

Does this mean if I have to use `sql` as input then we can't do unless we use it as processor? or are we looking to have `sql` separately as input in future?

**Comment by Jeffail** [comment]
Hey @uknow2009, right now you can use inputs that are scheduled like the [bloblang input](https://www.benthos.dev/docs/components/inputs/bloblang) in order to trigger processors on a regular interval. For example, something like:

```yaml
input:
  bloblang:
    interval: 1s
    mapping: 'root = "ignore"'
  processors:
    - sql:
        driver: postgresql
        result_codec: json_array
        data_source_name: postgres://foouser:foopass@localhost:5432/testdb?sslmode=disable
        query: "SELECT * FROM footable WHERE user_id = ?;"
        args: [ "blah" ]
```

**Comment by Jeffail** [comment]
Thanks everyone for the feedback. I'm closing this for now and will be putting a preliminary roadmap up for discussion "soon".

**Comment by gedw99** [comment]
:)
